Sources/AWSSDKSwift/Services/S3/S3+multipart_API.swift

```diff
@@ -251,3 +212,55 @@ public extension S3 {
         }
     }
 }
+
+
+extension S3 {
+    /// used internally in multipartUpload, loads all the parts once the multipart upload has been initiated
+    func multipartUploadParts(_ input: CreateMultipartUploadRequest, uploadId: String, on eventLoop: EventLoop, inputStream: @escaping () -> EventLoopFuture<Data?>) -> Future<[S3.CompletedPart]> {
+        let promise = eventLoop.makePromise(of: [S3.CompletedPart].self)
+        var completedParts: [S3.CompletedPart] = []
+
+        // function uploading part of a file and queueing up upload of the next part
+        func multipartUploadPart(partNumber: Int, uploadId: String, body: Data? = nil) {
+            // create upload data future, if there is no data to load because this is the first time this is called create a succeeded future
+            let uploadResult : Future<[S3.CompletedPart]>
+            if let body = body {
+                let request = S3.UploadPartRequest(
+                    body: body,
+                    bucket: input.bucket,
+                    contentLength: Int64(body.count),
+                    key: input.key,
+                    partNumber: partNumber,
+                    requestPayer: input.requestPayer,
+                    sSECustomerAlgorithm: input.sSECustomerAlgorithm,
+                    sSECustomerKey: input.sSECustomerKey,
+                    sSECustomerKeyMD5: input.sSECustomerKeyMD5,
+                    uploadId: uploadId
+                )
+                // request upload future
+                uploadResult = self.uploadPart(request).hop(to: eventLoop).map { output -> [S3.CompletedPart] in
+                    let part = S3.CompletedPart(eTag: output.eTag, partNumber: partNumber)
+                    completedParts.append(part)
+                    return completedParts
+                }
+            } else {
+                uploadResult = eventLoop.makeSucceededFuture([])
```

Comment:
As far as I can see, we will never upload anything for part 1. When this function is called for the very first time we haven't read any data yet. Therefore `inputStream` in the next round is called.

Reply:
The idea is we have the uploads and reads running in parallel. The first time you enter the function you can only run an inputStream. The second time you are uploading the data from the inputStream and running the second inputStream in parallel and so on, until the last inputStream returns that there is no data which also happens as the same time as the last upload. Will add comment detail this.

Although I have just noticed we are running this all on the same EventLoop so none of it happens in parallel.

Reply:
Except I'm talking crap they do run on separate eventloops as the AWSClient has its own eventloopgroup

Reply:
> Although I have just noticed we are running this all on the same EventLoop so none of it happens in parallel.

Of course this is running in parallel, even if it is running on the same EventLoop. That's the power of async programming.